doc/erofs.md

```diff
@@ -0,0 +1,431 @@
+the "suffix" that's left after removing the prefix.  The seed is
+`EROFS_XATTR_FILTER_SEED` plus the prefix index.  The lower 5 bits of the hash
+value (0..31) are used to determine which bit is used.
+
+For some reason a bit value of 1 here indicates the absence of a particular
+xattr, which is opposite to the usual arrangement.  You'd think it was for
+compatibility, but the filter is only engaged if the feature bit is present in
+the superblock.
+
+This feature got added in kernel commits:
+*   `3f339920175c` ("erofs: update on-disk format for xattr name filter")
+*   `fd73a4395d47` ("erofs: boost negative xattr lookup with bloom filter")
+
+### Data layout
+
+erofs has a bunch of different ways to represent the actual content associated
+with an inode (regular file content, directory entries, symlink target).
+
+We describe three of them here:
+*   plain
+*   inline
+*   chunked
+
+The data layout is chosen using some of the bits of the `format` field in the
+inode header.
+
+#### `EROFS_INODE_FLAT_PLAIN`
+
+In this case there's never any inline data.  The inode content is stored
+entirely as a series of contiguous blocks.  The offset of the first block is
+what goes in the `i_u` field (measured in blocks, not bytes).
+
+The number of blocks is determined by the `.size` field (divided by block size,
+rounded up).
+
+If the content is not a multiple of the blocksize then the last block should be
+0-padded.
+
+#### `EROFS_INODE_FLAT_INLINE`
+
+This is similar to `EROFS_INODE_FLAT_PLAIN` except if the content is not a
+multiple of the blocksize.  In that case, instead of 0-padding the last block
+to fill up a block, the content of the last block is stored directly inline
+with the inode, without padding.
+
+So, imagining the content is 2.5 blocks worth of data:
+*   the first block is the one pointed to by `i_u`
+*   the second block is the one immediately following it
+*   the last block is stored at the end of the inode
+
+The number of blocks is determined by the `.size` field, divided by block size,
+rounded down.  The remainder is the number of bytes of inline data.
+
+The inline data must be written in such a way that it does not cross a block
+boundary.  It is theoretically permitted for the inline data to be in a
+separate block (ie: the block directly following the inode data).  It is also
+permitted for the inode data itself to cross block boundaries.  There are a
+couple of caveats to be aware of, however:
+*   the alignment of inodes is 32 bytes, but the size of an extended inode is 64
+    bytes.  `mkfs.erofs` tries to ensure that extended inodes headers land
+    entirely within on disk block (for efficiency), but this isn't required by
+    the kernel.
+*   `mkfs.erofs` also tries to ensure that the inline data ends in the same
+    disk block as the last byte of the inode metadata (ie: inode header plus
+    xattrs).  This is theoretically not required by the kernel.
```

Comment:
yes, but the whole concept of inlined data is to save extra i/o as well as storage. Although the kernel supports it but it's never a design goal of `tail inline` inodes...

Reply:
I've been thinking about this IO amplification question a bit.  I have basically zero understanding of how filesystems actually work in the kernel, so this is possibly quite naive.

Maybe thinking about how much IO you need to do to load a single inode is a bit of a skewed metric.  I imagine that (during boot, for example) more or less the entire inode table is going to end up paged in and cached.  And in that case the best way to reduce IO would be to reduce the number of pages that need to be loaded, by using the least amount of padding possible, even if it hurts the alignment of a single inode. 

Do you have benchmarks or anything like that in this direction?

Reply:
> I've been thinking about this IO amplification question a bit. I have basically zero understanding of how filesystems actually work in the kernel, so this is possibly quite naive.
> 
> Maybe thinking about how much IO you need to do to load a single inode is a bit of a skewed metric. I imagine that (during boot, for example) more or less the entire inode table is going to end up paged in and cached. And in that case the best way to reduce IO would be to reduce the number of pages that need to be loaded, by using the least amount of padding possible, even if it hurts the alignment of a single inode.
> 
> Do you have benchmarks or anything like that in this direction?

In principle, the kernel doesn't do any policy strategies, because it might be good for your workload, but not good for others (e.g. users just would like to read specific inodes).  So you might win for some cases, but lose for other useful cases.  It's not quite something to serve for generic use cases.

EROFS always reads needed block-sized metadata as other kernel fses.  if you need to warm up metadata, you could use fadvise in advance.